### PR TITLE
Fix reading SQL_C_WCHAR on Mac

### PR DESCRIPTION
**Pull request by arcusfelis**
Includes:
- fix travis, update erlang versions
- one function to get "long" data (that has no known length).
- allow to pass varbinary to encoder.

Minuses:
- could be a bit slower, because we use SQLGetData in all the cases, even if data length is fixed.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -28,6 +25,7 @@ matrix:
       otp_release: 19.3
     - os: linux
       otp_release: 20.0
-    - os: osx
-      language: generic
-      script: ./rebar3 compile # Check that it at least compiles
+    - os: linux
+      otp_release: 21.0
+    - os: linux
+      otp_release: 22.0
```
Comment (NelsonVides) [inline .travis.yml:31]: Could also add OTP23

File: c_src/odbcserver.c [hunk 2]
```diff
@@ -3012,3 +2843,93 @@ static void str_tolower(char *str, int len)
...
+//      if (rc == SQL_NO_DATA) break;
+        if (!sql_success(rc)) break;
+
+        diagnos = get_diagnos(SQL_HANDLE_STMT, statement_handle(state), extended_errors(state));
+        truncated = (strcmp((char *)diagnos.sqlState, TRUNCATED) == 0);
+
+        // Number of written bytes without NULL terminator
+        got_bytes = (byte_len_or_ind >= chunk_bytes) ? (chunk_bytes - term_bytes) : byte_len_or_ind;
+        if (byte_len_or_ind == SQL_NULL_DATA) got_bytes = 0;
+        offset_bytes += got_bytes;
```
Comment (NelsonVides) [inline c_src/odbcserver.c]: This is confusing to quick-read, I'd do something like
```suggestion
        if (byte_len_or_ind == SQL_NULL_DATA)
            got_bytes = 0;
        offset_bytes += got_bytes;
```
or
```suggestion
        if (byte_len_or_ind == SQL_NULL_DATA) got_bytes = 0;
        
        offset_bytes += got_bytes;
```

File: c_src/odbcserver.c [hunk 3]
```diff
@@ -932,7 +933,8 @@ static db_result_msg db_param_query(byte *buffer, db_state *state)
 	    DO_EXIT(EXIT_FREE);
 	} 
     } else {
-	msg = encode_atom_message("param_badarg");
+        // params == NULL
+	    msg = encode_atom_message("param_badarg");
```
Comment (NelsonVides) [inline c_src/odbcserver.c]: Indentation is killing me, but anyway, what that's `params == NULL` comment?

File: c_src/odbcserver.c [hunk 4]
```diff
@@ -1358,8 +1361,14 @@ static db_result_msg encode_column_name_list(SQLSMALLINT num_of_columns,
...
 	
+		if (columns(state)[i].type.c == SQL_C_WCHAR) {
+		    /* retrived later by retrive_long_data */
+        } else
+		if (columns(state)[i].type.c == SQL_C_CHAR) {
+		    /* retrived later by retrive_long_data */
+        } else
 		if (columns(state)[i].type.c == SQL_C_BINARY) {
-		    /* retrived later by retrive_binary_data */
+		    /* retrived later by retrive_long_data */
```
Comment (NelsonVides) [inline c_src/odbcserver.c]: Man please fix that indentation, it's confusing as heck to read

File: c_src/odbcserver.c (hunk 3 above)
Comment (arcusfelis) [reply c_src/odbcserver.c]: probably remaining from some experiments.

File: c_src/odbcserver.c [hunk 5]
```diff
@@ -1544,50 +1518,39 @@ static void encode_column_dyn(db_column column, int column_nr,
...
-            retrive_long_data(column, column_nr, SQL_C_CHAR, state, &bufferptr, &result_len);
-
-			if binary_strings(state) {
-				 ei_x_encode_binary(&dynamic_buffer(state), bufferptr, result_len);
-			} else {
-				ei_x_encode_string(&dynamic_buffer(state), bufferptr);
-			}
 
+    case SQL_C_CHAR: {
+            char *bufferptr = 0;
```
Comment (NelsonVides) [inline c_src/odbcserver.c:1523]: I know it means the same shit, but better follow conventions
```suggestion
            char *bufferptr = NULL;
```

File: c_src/odbcserver.c [hunk 6]
```diff
@@ -1544,50 +1518,39 @@ static void encode_column_dyn(db_column column, int column_nr,
...
+            } else {
+                ei_x_encode_atom(&dynamic_buffer(state), "null");
+            }
+        break; }
+
+    case SQL_C_WCHAR: {
+        // Read chunks of data.
+        // There is pretty high change to get <<0,0,0...>> as a column data on some systems otherwise
+        // https://docs.microsoft.com/en-us/sql/relational-databases/native-client/features/odbc-driver-behavior-change-when-handling-character-conversions?view=sql-server-ver15
+        char *bufferptr = 0;
```
Comment (NelsonVides) [inline c_src/odbcserver.c:1542]: As above

File: c_src/odbcserver.c [hunk 7]
```diff
@@ -1544,50 +1518,39 @@ static void encode_column_dyn(db_column column, int column_nr,
...
+        int result_len = 0; /* in bytes */
+        const int sizeof_element = sizeof(SQLWCHAR); // 2 bytes
+        retrive_long_data(column, column_nr, SQL_C_WCHAR, &bufferptr, &result_len, sizeof_element, sizeof_element, state);
+        if (bufferptr) {
+            ei_x_encode_binary(&dynamic_buffer(state), bufferptr, result_len);
             free(bufferptr);
+        } else {
+            ei_x_encode_atom(&dynamic_buffer(state), "null");
         }
```
Comment (NelsonVides) [inline c_src/odbcserver.c:1551]: This same logic is repeated in three places, any chance to refactor it?

**Pull request closed without merging**